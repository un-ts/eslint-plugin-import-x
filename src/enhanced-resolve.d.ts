export type AliasOptionNewRequest = string | false | string[]

export type AliasOptions = {
  [index: string]: AliasOptionNewRequest
}

export type ExtensionAliasOptions = {
  [index: string]: string | string[]
}

export type ResolveOptions = {
  /**
   * A list of module alias configurations or an object which maps key to value
   */
  alias?: AliasOptions | AliasOption[]

  /**
   * A list of module alias configurations or an object which maps key to value, applied only after modules option
   */
  fallback?: AliasOptions | AliasOption[]

  /**
   * An object which maps extension to extension aliases
   */
  extensionAlias?: ExtensionAliasOptions

  /**
   * A list of alias fields in description files
   */
  aliasFields?: Array<string | string[]>

  /**
   * Whether or not the unsafeCache should include request context as part of the cache key.
   */
  cacheWithContext?: boolean

  /**
   * A list of description files to read from
   */
  descriptionFiles?: string[]

  /**
   * A list of exports field condition names.
   */
  conditionNames?: string[]

  /**
   * Enforce that a extension from extensions must be used
   */
  enforceExtension?: boolean

  /**
   * A list of exports fields in description files
   */
  exportsFields?: Array<string | string[]>

  /**
   * A list of imports fields in description files
   */
  importsFields?: Array<string | string[]>

  /**
   * A list of extensions which should be tried for files
   */
  extensions?: string[]

  /**
   * Use this cache object to unsafely cache the successful requests
   */
  unsafeCache?: boolean | object

  /**
   * Resolve symlinks to their symlinked location
   */
  symlinks?: boolean

  /**
   * A list of directories to resolve modules from, can be absolute path or folder name
   */
  modules?: string | string[]

  /**
   * A list of main fields in description files
   */
  mainFields?: Array<
    string | string[] | { name: string | string[]; forceRelative: boolean }
  >

  /**
   * A list of main files in directories
   */
  mainFiles?: string[]

  /**
   * A PnP API that should be used - null is "never", undefined is "auto"
   */
  pnpApi?: null | PnpApi

  /**
   * A list of root paths
   */
  roots?: string[]

  /**
   * The request is already fully specified and no extensions or directories are resolved for it
   */
  fullySpecified?: boolean

  /**
   * Resolve to a context instead of a file
   */
  resolveToContext?: boolean

  /**
   * A list of resolve restrictions
   */
  restrictions?: Array<string | RegExp>

  /**
   * Use only the sync constraints of the file system calls
   */
  useSyncFileSystemCalls?: boolean

  /**
   * Prefer to resolve module requests as relative requests before falling back to modules
   */
  preferRelative?: boolean

  /**
   * Prefer to resolve server-relative urls as absolute paths before falling back to resolve in roots
   */
  preferAbsolute?: boolean
}
