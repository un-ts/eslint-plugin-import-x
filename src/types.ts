import type { TSESLint, TSESTree } from '@typescript-eslint/utils'
import type { MinimatchOptions } from 'minimatch'
import type { NapiResolveOptions as ResolveOptions } from 'oxc-resolver'
import type { KebabCase } from 'type-fest'

import type { ImportType as ImportType_, PluginName } from './utils'
import type {
  LegacyImportResolver,
  LegacyResolver,
} from './utils/legacy-resolver-settings'

export type {
  LegacyResolver,
  // ResolverName
  LegacyResolverName,
  LegacyResolverName as ResolverName,
  // ImportResolver
  LegacyImportResolver,
  LegacyImportResolver as ImportResolver,
  // ResolverResolve
  LegacyResolverResolve,
  LegacyResolverResolve as ResolverResolve,
  // ResolverResolveImport
  LegacyResolverResolveImport,
  LegacyResolverResolveImport as ResolverResolveImport,
  // ResolverRecord
  LegacyResolverRecord,
  LegacyResolverRecord as ResolverRecord,
  // ResolverObject
  LegacyResolverObject,
  LegacyResolverObject as ResolverObject,
} from './utils/legacy-resolver-settings'

export type ImportType = ImportType_ | 'object' | 'type'

export type NodeResolverOptions = {
  extensions?: readonly string[]
  moduleDirectory?: string[]
  paths?: string[]
}

export type WebpackResolverOptions = {
  config?: string | { resolve: ResolveOptions }
  'config-index'?: number
  env?: Record<string, unknown>
  argv?: Record<string, unknown>
}

export type TsResolverOptions = {
  alwaysTryTypes?: boolean
  project?: string[] | string
  extensions?: string[]
} & ResolveOptions

// TODO: remove prefix New in the next major version
export type NewResolverResolve = (
  modulePath: string,
  sourceFile: string,
) => ResolvedResult

// TODO: remove prefix New in the next major version
export type NewResolver = {
  interfaceVersion: 3
  /** optional name for the resolver, this is used in logs/debug output */
  name?: string
  resolve: NewResolverResolve
}

export type FileExtension = `.${string}`

export type DocStyle = 'jsdoc' | 'tomdoc'

export type Arrayable<T> = T | readonly T[]

export type ResultNotFound = {
  found: false
  path?: undefined
}

export type ResultFound = {
  found: true
  path: string | null
}

export type Resolver = LegacyResolver | NewResolver

export type ResolvedResult = ResultNotFound | ResultFound

export type ImportSettings = {
  cache?: {
    lifetime?: number | 'âˆž' | 'Infinity'
  }
  coreModules?: string[]
  docstyle?: DocStyle[]
  extensions?: readonly FileExtension[]
  externalModuleFolders?: string[]
  ignore?: string[]
  internalRegex?: string
  parsers?: Record<string, readonly FileExtension[]>
  resolve?: NodeResolverOptions
  resolver?: LegacyImportResolver
  'resolver-legacy'?: LegacyImportResolver
  'resolver-next'?: NewResolver[]
}

export type WithPluginName<T extends string | object> = T extends string
  ? `${PluginName}/${KebabCase<T>}`
  : {
      [K in keyof T as WithPluginName<`${KebabCase<K & string>}`>]: T[K]
    }

export type PluginSettings = WithPluginName<ImportSettings>

export type PluginConfig = {
  plugins?: [PluginName]
  settings?: PluginSettings
  rules?: Record<`${PluginName}/${string}`, TSESLint.Linter.RuleEntry>
} & TSESLint.Linter.ConfigType

export type PluginFlatBaseConfig = {
  settings?: PluginSettings
  rules?: Record<`${PluginName}/${string}`, TSESLint.FlatConfig.RuleEntry>
} & TSESLint.FlatConfig.Config

export type PluginFlatConfig = PluginFlatBaseConfig & {
  name?: `${PluginName}/${string}`
}

export type RuleContext<
  TMessageIds extends string = string,
  TOptions extends readonly unknown[] = readonly unknown[],
> = Readonly<{
  languageOptions?: TSESLint.FlatConfig.LanguageOptions
  settings: PluginSettings
}> &
  Omit<TSESLint.RuleContext<TMessageIds, TOptions>, 'settings'>

export type ChildContext = {
  cacheKey: string
  settings: PluginSettings
  parserPath?: string | null
  parserOptions?: TSESLint.ParserOptions
  languageOptions?: TSESLint.FlatConfig.LanguageOptions
  path: string
  filename?: string
}

export type ParseError = {
  lineNumber: number
  column: number
} & Error

export type CustomESTreeNode<
  Type extends string,
  T extends object = object,
> = Omit<TSESTree.BaseNode, 'type'> & {
  type: Type
} & T

export type ExportDefaultSpecifier = CustomESTreeNode<'ExportDefaultSpecifier'>

export type ExportNamespaceSpecifier = CustomESTreeNode<
  'ExportNamespaceSpecifier',
  { exported: TSESTree.Identifier }
>

export type PathGroup = {
  pattern: string
  group: ImportType
  patternOptions?: MinimatchOptions
  position?: 'before' | 'after'
}

export type AlphabetizeOptions = {
  caseInsensitive: boolean
  order: 'ignore' | 'asc' | 'desc'
  orderImportKind: 'ignore' | 'asc' | 'desc'
}
