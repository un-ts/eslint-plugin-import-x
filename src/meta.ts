import { createRequire } from 'node:module'

import type { CjsRequire } from '@pkgr/core'

const cjsRequire: CjsRequire =
  typeof require === 'undefined'
    ? createRequire(import.meta.url)
    : /* istanbul ignore next */ require

export const { name, version } = cjsRequire<{ name: string; version: string }>(
  '../package.json',
)

export const meta = { name, version }
