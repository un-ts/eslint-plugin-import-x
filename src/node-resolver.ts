import { isBuiltin } from 'node:module'
import path from 'node:path'

import { ResolverFactory } from 'oxc-resolver'
import type { NapiResolveOptions as ResolveOptions } from 'oxc-resolver'

import type { NewResolver } from './types'

export type NodeResolverOptions = {
  /**
   * Attempt to resolve these extensions in order.
   * If multiple files share the same name but have different extensions,
   * will resolve the one with the extension listed first in the array and skip the rest.
   *
   * @default ['.mjs', '.cjs', '.js', '.json', '.node']
   */
  extensions?: string[]
  /**
   * Condition names for exports field which defines entry points of a package.
   * The key order in the exports field is significant. During condition matching, earlier entries have higher priority and take precedence over later entries.
   *
   * @default ['import', 'require', 'default']
   */
  conditionNames?: string[]
  /**
   * A list of main fields in description files
   *
   * @default ['module', 'main']
   */
  mainFields?: string[]
} & ResolveOptions

export function createNodeResolver({
  extensions = ['.mjs', '.cjs', '.js', '.json', '.node'],
  conditionNames = ['import', 'require', 'default'],
  mainFields = ['module', 'main'],
  ...restOptions
}: NodeResolverOptions = {}): NewResolver {
  const resolver = new ResolverFactory({
    extensions,
    conditionNames,
    mainFields,
    ...restOptions,
  })

  // shared context across all resolve calls

  return {
    interfaceVersion: 3,
    name: 'eslint-plugin-import-x built-in node resolver',
    resolve(modulePath, sourceFile) {
      if (isBuiltin(modulePath)) {
        return { found: true, path: null }
      }

      if (modulePath.startsWith('data:')) {
        return { found: true, path: null }
      }

      try {
        const resolved = resolver.sync(path.dirname(sourceFile), modulePath)
        if (resolved.path) {
          return { found: true, path: resolved.path }
        }
        return { found: false }
      } catch {
        return { found: false }
      }
    },
  }
}
