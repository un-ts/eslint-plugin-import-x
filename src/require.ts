import { createRequire } from 'node:module'

import type { CjsRequire } from './types.js'

const importMetaUrl = import.meta.url

export const cjsRequire: CjsRequire = importMetaUrl
  ? createRequire(importMetaUrl)
  : /* istanbul ignore next */ require
