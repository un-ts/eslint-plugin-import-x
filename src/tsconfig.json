{
  "extends": "../tsconfig.base",
  "compilerOptions": {
    "rootDir": ".",
    "outDir": "../lib"
  },
  "include": ["."]
}
