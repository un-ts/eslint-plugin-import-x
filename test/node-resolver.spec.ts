import path from 'node:path'

import { createNodeResolver } from '../src/node-resolver'

const resolver = createNodeResolver()

function expectResolve(source: string, expected: boolean | string) {
  it(`${source} => ${expected}`, () => {
    let requireResolve: string | undefined
    try {
      requireResolve = require.resolve(source, { paths: [__dirname] })
    } catch {
      // ignore
    }

    expect({
      source,
      expected,
      requireResolve,
    }).toMatchSnapshot()

    const result = resolver.resolve(source, __filename)
    expect({
      source,
      expected,
      result,
    }).toMatchSnapshot()

    if (typeof expected === 'string') {
      expect(result.path).toBe(path.resolve(expected))
    } else {
      expect(result.found).toBe(expected)
    }
  })
}

describe('builtin', () => {
  expectResolve('path', true)
  expectResolve('node:path', true)
})

describe('modules', () => {
  expectResolve('jest', true)
  expectResolve('@sukka/does-not-exists', false)
})

describe('relative', () => {
  expectResolve('../package.json', 'package.json')
  expectResolve('../.github/dependabot.yml', false)
  expectResolve('../babel.config.js', 'babel.config.js')
  expectResolve('../test/index.js', 'test/index.js')
  expectResolve('../test/', 'test/index.js')
  expectResolve('../test', 'test/index.js')

  expectResolve('../inexistent.js', false)
})
