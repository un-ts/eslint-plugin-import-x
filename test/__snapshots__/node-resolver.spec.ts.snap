// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builtin node:path => true 1`] = `
{
  "expected": true,
  "requireResolve": "node:path",
  "source": "node:path",
}
`;

exports[`builtin node:path => true 2`] = `
{
  "expected": true,
  "result": {
    "found": true,
    "path": null,
  },
  "source": "node:path",
}
`;

exports[`builtin path => true 1`] = `
{
  "expected": true,
  "requireResolve": "path",
  "source": "path",
}
`;

exports[`builtin path => true 2`] = `
{
  "expected": true,
  "result": {
    "found": true,
    "path": null,
  },
  "source": "path",
}
`;

exports[`modules @sukka/does-not-exists => false 1`] = `
{
  "expected": false,
  "requireResolve": undefined,
  "source": "@sukka/does-not-exists",
}
`;

exports[`modules @sukka/does-not-exists => false 2`] = `
{
  "expected": false,
  "result": {
    "found": false,
  },
  "source": "@sukka/does-not-exists",
}
`;

exports[`modules jest => true 1`] = `
{
  "expected": true,
  "requireResolve": "<ROOT>/node_modules/jest/build/index.js",
  "source": "jest",
}
`;

exports[`modules jest => true 2`] = `
{
  "expected": true,
  "result": {
    "found": true,
    "path": "<ROOT>/node_modules/jest/build/index.js",
  },
  "source": "jest",
}
`;

exports[`relative ../.github/dependabot.yml => false 1`] = `
{
  "expected": false,
  "requireResolve": undefined,
  "source": "../.github/dependabot.yml",
}
`;

exports[`relative ../.github/dependabot.yml => false 2`] = `
{
  "expected": false,
  "result": {
    "found": false,
  },
  "source": "../.github/dependabot.yml",
}
`;

exports[`relative ../babel.config.js => babel.config.js 1`] = `
{
  "expected": "babel.config.js",
  "requireResolve": "<ROOT>/babel.config.js",
  "source": "../babel.config.js",
}
`;

exports[`relative ../babel.config.js => babel.config.js 2`] = `
{
  "expected": "babel.config.js",
  "result": {
    "found": true,
    "path": "<ROOT>/babel.config.js",
  },
  "source": "../babel.config.js",
}
`;

exports[`relative ../inexistent.js => false 1`] = `
{
  "expected": false,
  "requireResolve": undefined,
  "source": "../inexistent.js",
}
`;

exports[`relative ../inexistent.js => false 2`] = `
{
  "expected": false,
  "result": {
    "found": false,
  },
  "source": "../inexistent.js",
}
`;

exports[`relative ../package.json => package.json 1`] = `
{
  "expected": "package.json",
  "requireResolve": "<ROOT>/package.json",
  "source": "../package.json",
}
`;

exports[`relative ../package.json => package.json 2`] = `
{
  "expected": "package.json",
  "result": {
    "found": true,
    "path": "<ROOT>/package.json",
  },
  "source": "../package.json",
}
`;

exports[`relative ../test => test/index.js 1`] = `
{
  "expected": "test/index.js",
  "requireResolve": "<ROOT>/test/index.js",
  "source": "../test",
}
`;

exports[`relative ../test => test/index.js 2`] = `
{
  "expected": "test/index.js",
  "result": {
    "found": true,
    "path": "<ROOT>/test/index.js",
  },
  "source": "../test",
}
`;

exports[`relative ../test/ => test/index.js 1`] = `
{
  "expected": "test/index.js",
  "requireResolve": "<ROOT>/test/index.js",
  "source": "../test/",
}
`;

exports[`relative ../test/ => test/index.js 2`] = `
{
  "expected": "test/index.js",
  "result": {
    "found": true,
    "path": "<ROOT>/test/index.js",
  },
  "source": "../test/",
}
`;

exports[`relative ../test/index.js => test/index.js 1`] = `
{
  "expected": "test/index.js",
  "requireResolve": "<ROOT>/test/index.js",
  "source": "../test/index.js",
}
`;

exports[`relative ../test/index.js => test/index.js 2`] = `
{
  "expected": "test/index.js",
  "result": {
    "found": true,
    "path": "<ROOT>/test/index.js",
  },
  "source": "../test/index.js",
}
`;
